<div style="font-size:15pt;margin-bottom:0;text-align: right"><b>FIT1043 Introduction to Data Science</b></div>
<div style="font-size:15pt;margin-top:0;text-align: right"><b>Assignment 2</b></div>
<br>
<div style="text-align: right"> Loh Jing Wei </div>
<div style="text-align: right"> 30856183 </div>
<div style="text-align: right"> <i> 29th April 2022 </i> </div>
<hr style="border:0.5px solid gray"> </hr>

***
## 1. Introduction

This assignment involves using data about essays from a CSV file named FIT1043-Essay-Features.csv to create a model which will predict the scores of the essays. <br>
The data we will be dealing with is labelled data, hence, supervised machine learning method will be used. <br>
The label for this dataset will be the score of the essays which is a categorical-ordinal data type, meanwhile features will be chosen using filter selection method. After choosing our features, we split the dataset into testing and training datasets. <br>
We will normalize and scale the data using StandardScaler. Subsequently, Support vector machine (specifically SVC) is chosen as the main algorithm to form our model. <br>
Confusion matrix, accuracy score and Quadratic Weighted Kappa (QWK) are used to assess our model. <br>
The model will then be used to predict scores for dataset from another CSV file named FIT1043-Essay-Features-Submission.csv, and the result (predicted scores for the essays) will be exported into a CSV file called 30856183-Jing-Wei-Loh-1.csv. This CSV file will be uploaded to Kaggle.<br>


### Importing the necessary libraries and reading files

We import the library <b>pandas</b> to use the data structure DataFrame and the related functions to read data from CSV files as well as data manipulation and analysis. <br>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

<b>FIT1043-Essay-Features.csv</b> file is read and stored as DataFrame with name essay_feature. <br>
This FIT1043-Essay-Features.csv file describes features from a set of essays and provide information about the essays.<br>
<b>30856183-Jing-Wei-Loh-1.csv</b> file is read and stored as DataFrame named submission. It will be used to store the predicted score for Kaggle submission. <br>

In [2]:
essay_features = pd.read_csv('FIT1043-Essay-Features.csv')
submission = pd.read_csv('30856183-Jing-Wei-Loh-1.csv')

Check for the number of columns and rows of essay_features DataFrame

In [3]:
essay_features.shape

(1332, 19)

There are 1332 rows and 19 columns in total

Check for the first 5 entries of the dataframe

In [4]:
essay_features.head()

essayid  chars  words  commas  apostrophes  punctuations  avg_word_length  \
0     1457   2153    426      14            6             0         5.053991   
1      503   1480    292       9            7             0         5.068493   
2      253   3964    849      19           26             1         4.669022   
3      107    988    210       8            7             0         4.704762   
4     1450   3139    600      13            8             0         5.231667   

   sentences  questions  avg_word_sentence         POS  POS/total_words  \
0         16          0          26.625000  423.995272         0.995294   
1         11          0          26.545455  290.993103         0.996552   
2         49          2          17.326531  843.990544         0.994100   
3         12          0          17.500000  207.653784         0.988828   
4         24          1          25.000000  594.652150         0.991087   

   prompt_words  prompt_words/total_words  synonym_words  \
0           207                  0.485915            105   
1           148                  0.506849             77   
2           285                  0.335689            130   
3           112                  0.533333             62   
4           255                  0.425000            165   

   synonym_words/total_words  unstemmed  stemmed  score  
0                   0.246479        424      412      4  
1                   0.263699        356      345      4  
2                   0.153121        750      750      4  
3                   0.295238        217      209      3  
4                   0.275000        702      677      4

Use .describe() function to find the count, mean, standard deviation, minimum, 25th percentile, 50th percentile, 75th percentile and maximum value for each column

In [5]:
essay_features.describe()

essayid        chars        words       commas  apostrophes  \
count  1332.00000  1332.000000  1332.000000  1332.000000  1332.000000   
mean    905.27027  2101.745495   424.485736    14.667417     8.141141   
std     526.68760   865.963750   171.873730    10.920781     6.124520   
min       0.00000   169.000000    36.000000     0.000000     2.000000   
25%     442.75000  1527.250000   310.000000     7.000000     4.000000   
50%     914.50000  2029.500000   411.000000    13.000000     6.000000   
75%    1369.75000  2613.500000   525.000000    21.000000    11.000000   
max    1799.00000  6142.000000  1170.000000    72.000000    51.000000   

       punctuations  avg_word_length    sentences    questions  \
count    1332.00000      1332.000000  1332.000000  1332.000000   
mean        0.47973         4.939762    19.704204     1.222973   
std         1.27168         0.231071    19.202731     1.847446   
min         0.00000         2.231322     0.000000     0.000000   
25%         0.00000         4.791679    13.000000     0.000000   
50%         0.00000         4.946059    18.000000     1.000000   
75%         0.00000         5.092938    24.000000     2.000000   
max        26.00000         5.681429   642.000000    17.000000   

       avg_word_sentence          POS  POS/total_words  prompt_words  \
count        1332.000000  1332.000000      1332.000000   1332.000000   
mean           23.884687   420.596542         0.989935    198.913664   
std            11.160020   170.985111         0.007308     82.729266   
min             1.084112    35.647059         0.924771     14.000000   
25%            19.142857   305.406284         0.987758    144.000000   
50%            22.030331   406.982869         0.991572    193.000000   
75%            26.048234   520.739458         0.994425    246.000000   
max           303.000000  1158.984563         1.000000    669.000000   

       prompt_words/total_words  synonym_words  synonym_words/total_words  \
count               1332.000000     1332.00000                1332.000000   
mean                   0.469164      110.16967                   0.263846   
std                    0.052466       43.96192                   0.038870   
min                    0.288889       11.00000                   0.027299   
25%                    0.435709       81.00000                   0.238423   
50%                    0.465852      107.50000                   0.262872   
75%                    0.500000      134.00000                   0.288277   
max                    0.961207      355.00000                   0.465517   

         unstemmed      stemmed        score  
count  1332.000000  1332.000000  1332.000000  
mean    468.987988   455.507508     3.427177  
std     159.447449   155.751220     0.774275  
min      48.000000    50.000000     1.000000  
25%     361.000000   350.750000     3.000000  
50%     463.000000   448.000000     3.000000  
75%     581.000000   561.250000     4.000000  
max     750.000000   750.000000     6.000000

***
## 2. Supervised Learning

There are 2 type of machine learning approaches: 
- Supervised Machine learning 
- Unsupervised machine learning <br>

<b> Supervised machine learning </b> uses labelled data and the algorithms learn to predict the output from the input data. <br>
<b> Labelled data </b> is a group of samples that has labels which identify its properties or characteristics.
Labels can come from observations or asking specialists about the data. When new unlabelled data is given to the model, the model is able to predict a likely label for the data.

There are 2 type of problems which use supervised machine learning: Classification and Regression. <br>
- Classification is used when output variable is categorical data (which in our case, score is a categorical-ordinal data)
- Regression is used when output variable is a continuous value

### Feature Selection

<b>Feature selection</b> is used to reduce the number of input variables by choosing those that are relevant to the target variable (the input variables which have a correlation with the output variable), thus these input variables are useful for a model to predict the target variable.

There are several type of feature selection methods for supervised machine learning such as <b>Filter and Wrapper</b>.<br>
We will be using the filter selection method, which uses statistical measures to evaluate the relationship between each input variable and the target variable. <br>
Since our data has numerical input and categorical output, the technique we will be using is <b>Kendall’s rank coefficient</b>

We import the library <b>scipy.stats</b> to calculate Kendall’s tau<br>
The <b>kendalltau function</b> returns a value of -1 to 1. A value approaching -1 indicates that there is negative correlation and a value approaching +1 indicates a positive correlation.

In [6]:
import scipy.stats

In [7]:
scipy.stats.kendalltau(x=essay_features['essayid'],y=essay_features['score'])

KendalltauResult(correlation=0.020404014642527515, pvalue=0.3365365839873953)

In [8]:
scipy.stats.kendalltau(x=essay_features['chars'],y=essay_features['score'])

KendalltauResult(correlation=0.5662913993383564, pvalue=1.12074440073025e-156)

In [9]:
scipy.stats.kendalltau(x=essay_features['words'],y=essay_features['score'])

KendalltauResult(correlation=0.5455409604156707, pvalue=2.282631844331847e-145)

In [10]:
scipy.stats.kendalltau(x=essay_features['commas'],y=essay_features['score'])

KendalltauResult(correlation=0.44986484334829663, pvalue=8.888516380233097e-97)

In [11]:
scipy.stats.kendalltau(x=essay_features['apostrophes'],y=essay_features['score'])

KendalltauResult(correlation=0.2799489001226123, pvalue=3.974133761586599e-37)

In [12]:
scipy.stats.kendalltau(x=essay_features['punctuations'],y=essay_features['score'])

KendalltauResult(correlation=0.19851964868501998, pvalue=2.21907699545917e-15)

In [13]:
scipy.stats.kendalltau(x=essay_features['avg_word_length'],y=essay_features['score'])

KendalltauResult(correlation=0.2455979911658759, pvalue=6.028923704303055e-31)

In [14]:
scipy.stats.kendalltau(x=essay_features['sentences'],y=essay_features['score'])

KendalltauResult(correlation=0.49154562216707187, pvalue=8.30502458081386e-115)

In [15]:
scipy.stats.kendalltau(x=essay_features['questions'],y=essay_features['score'])

KendalltauResult(correlation=0.2496394645307749, pvalue=1.1720789275881243e-25)

In [16]:
scipy.stats.kendalltau(x=essay_features['avg_word_sentence'],y=essay_features['score'])

KendalltauResult(correlation=-0.02071340101424389, pvalue=0.3293877474131538)

In [17]:
scipy.stats.kendalltau(x=essay_features['POS'],y=essay_features['score'])

KendalltauResult(correlation=0.5461931394880581, pvalue=6.338625439016613e-146)

In [18]:
scipy.stats.kendalltau(x=essay_features['POS/total_words'],y=essay_features['score'])

KendalltauResult(correlation=0.20790478624388226, pvalue=1.2204822056438548e-22)

In [19]:
scipy.stats.kendalltau(x=essay_features['prompt_words'],y=essay_features['score'])

KendalltauResult(correlation=0.541249699102086, pvalue=7.627491789118741e-143)

In [20]:
scipy.stats.kendalltau(x=essay_features['prompt_words/total_words'],y=essay_features['score'])

KendalltauResult(correlation=0.04172987733066745, pvalue=0.04938961379675058)

In [21]:
scipy.stats.kendalltau(x=essay_features['synonym_words'],y=essay_features['score'])

KendalltauResult(correlation=0.46910891005258765, pvalue=1.9308321900552002e-107)

In [22]:
scipy.stats.kendalltau(x=essay_features['synonym_words/total_words'],y=essay_features['score'])

KendalltauResult(correlation=-0.23775364489280706, pvalue=4.214515958724073e-29)

In [23]:
scipy.stats.kendalltau(x=essay_features['unstemmed'],y=essay_features['score'])

KendalltauResult(correlation=0.5555578679361636, pvalue=4.107544572054594e-150)

In [24]:
scipy.stats.kendalltau(x=essay_features['stemmed'],y=essay_features['score'])

KendalltauResult(correlation=0.5573415908524881, pvalue=3.0493449550250838e-151)

We will take features with tau value above 0.4

### Separate the features and the label (Score)

Since we are taking the features with tau value above 0.4, the features chosen are as given below. <br>
Features: 'chars', 'words', 'commas', 'sentences', 'POS', 'prompt_words', 'synonym_words', 'unstemmed', 'stemmed' <br>
Label: ‘score’

In [25]:
X = essay_features.iloc[:,[1,2,3,7,10,12,16,17]].values
y = essay_features.iloc[:, 18].values 

### Split dataset into training and test data set

We import <b>train_test_split</b> function from the sklearn library, and we will be using this function to split the data into training and testing data set. <br>
The training data set will be used to train our machine learning models and the testing data set will be given to the model to make predictions. <br>
We will split the data into 80% training and 20% testing data set.

In [26]:
# Split dataset into training and test data set
# Take a portion of the data or take from a “random” sample of the data
# 80% for the training data and 20% for the testing data
from sklearn.model_selection import train_test_split 
X_train, X_test, y_train, y_test = train_test_split(
X, y, test_size = 0.2, random_state = 0 
)

***
## 3. Classification

### Binary and multi-class classification

<b>Binary-class Classification</b> refers to tasks with only 2 two class labels. For example thief detection (thief or not). <br>
<b>Multi-class Classification</b> refers to tasks with more than 2 class labels. For example classification of music genre. <br>
https://machinelearningmastery.com/types-of-classification-in-machine-learning/

### Data Scaling / Normalising

<b>Purpose of normalising / scaling of data</b><br>
The reason why we normalise and scale our data is that machine learning algorithm works faster and better when features (input variables) have similar scale and are close to being normally distributed. <br> 
- <b>Scaling</b> refers to changing the range of the data so that the variables/features fit within a specific range (for example 0-1) and the shape remain unchanged. As a result, we are able to compare different variables.<br>
- <b>Normalisation</b> changes your data so that it is normally distributed. It is needed when we use machine learning which assumes data to be normally distributed. <br>

https://towardsdatascience.com/scale-standardize-or-normalize-with-scikit-learn-6ccc7d176a02

In our case, we will be using SVM which does not assume a normally dsitributed data. Hence we will be doing <b>Scaling</b>.

<b>Ways to do scaling </b><br>
There are different way to do scaling, namely MinMaxScaler, RobustScaler and StandardScaler.<br>
- <b>StandardScaler</b> results in distribution with standard deviation = variance = 1, and mean of distribution = 0. - <br>
- <b>MinMaxScaler</b> is used when there is no outliers to reduce and results in distribution with range 0 to 1. <br>
- <b>RobustScaler</b> is used when there are outliers and there is a need to reduce the outliers' influence.<br>
After considering the options, the StandardScaler will be used as outliers cannot be confirmed. <br>
https://towardsdatascience.com/scale-standardize-or-normalize-with-scikit-learn-6ccc7d176a02

We import the library <b>sklearn.preprocessing</b> to use the function StandardScaler to scale our data

In [27]:
from sklearn.preprocessing import StandardScaler 
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

### Support Vector Machine

<b>Support vector machine (SVM) </b><br>
SVM is an algorithm used for supervised machine learning. <br>
SVM can be used as a classifier, support vector classification (SVC), or used as a regressor, support vector regression (SVR). <br>

<b>Support vector classification (SVC)</b> <br>
In SVC, the algorithm produces a line in an N-dimensional space (N = number of features), called hyperplane, between the data to separate them into different classes. For example, if there is 2 features the hyperplane is a line, if there is 3 features the hyperplane is a plane, etc. <br>
It uses support vectors (data points that are close to the hyperplane and have higher influence towards the orientation of the hyperplane) to build the SVM. <br>
Hyperplane is useful because it allows us to separate data that are not linearly separable. (By creating a new dimension/plane, the data becomes linearly separable, allowing for the optimal line to be drawn) <br>

<b>Support vector regression (SVR) and comparison to linear regression </b> <br>
In SVR, although this step is similar to linear regression in terms of using a line to regress data, the method is fundamentally different. While line regression minimizes error between predicted and actual values using the line of best fit, the hyperplane is fitted between a threshold of values called epsilon-insensitive tube, where only the points within the tube are considered. Because the points within the tube have the least error rate, this gives us a better fitting model.

https://towardsdatascience.com/support-vector-machine-introduction-to-machine-learning-algorithms-934a444fca47

<b>Kernel</b><br>
SVM kernel is a function which transforms the input data space (low number of features) into a higher dimensional space (high number of features). <br>
It acts as a shortcut to avoid complex calculations and allows us to form hyperplanes without raising the complexity of our model. <br>

By using kernels, we can solve non-linear problems using linear classifiers through the formation of hyperplanes. <br>
- Kernel function is used when data points cannot be linearly seperated. The SVM kernel solves this problem by introducing additional features (by adding new feature z (z=x^2+y^2) on existing x and y input features). <br>

The popular Kernel function includes: linear, polynomial, radial basis function and sigmoid.

https://techvidvan.com/tutorials/svm-kernel-functions/#:~:text=A%20kernel%20is%20a%20function,number%20of%20dimensions%20using%20kernels. <br>
https://www.analyticsvidhya.com/blog/2017/09/understaing-support-vector-machine-example-code/

In [28]:
from sklearn.svm import SVC
clf = SVC(decision_function_shape='ovo',kernel='linear')
clf.fit(X_train,y_train)

SVC(decision_function_shape='ovo', kernel='linear')

Predict the score using our model for the testing dataset

In [29]:
y_pred = clf.predict(X_test)

In [30]:
y_pred

array([4, 3, 3, 2, 3, 4, 4, 4, 2, 4, 4, 3, 4, 3, 3, 4, 3, 4, 4, 4, 2, 3,
       3, 3, 3, 4, 4, 4, 3, 4, 4, 3, 3, 4, 4, 3, 4, 4, 3, 3, 4, 2, 3, 3,
       3, 3, 3, 3, 2, 3, 4, 3, 4, 4, 4, 4, 4, 3, 3, 4, 3, 4, 3, 3, 4, 4,
       4, 4, 4, 4, 3, 3, 4, 4, 4, 3, 3, 4, 4, 3, 4, 3, 4, 3, 3, 4, 3, 3,
       4, 4, 3, 4, 4, 2, 4, 3, 3, 4, 3, 2, 4, 4, 3, 4, 4, 3, 4, 4, 3, 4,
       2, 3, 3, 3, 3, 4, 3, 4, 3, 3, 3, 3, 4, 4, 3, 3, 3, 4, 4, 4, 4, 4,
       4, 3, 4, 3, 3, 4, 4, 3, 3, 4, 4, 3, 3, 3, 3, 3, 4, 4, 4, 3, 4, 4,
       4, 3, 2, 3, 4, 3, 4, 4, 3, 3, 3, 4, 4, 4, 3, 3, 3, 3, 3, 3, 4, 4,
       4, 4, 4, 2, 3, 4, 3, 3, 3, 4, 3, 3, 3, 3, 3, 4, 3, 3, 2, 3, 4, 4,
       3, 4, 3, 4, 4, 3, 3, 4, 3, 3, 3, 3, 4, 4, 3, 3, 3, 3, 4, 4, 3, 3,
       4, 4, 4, 4, 4, 3, 2, 4, 3, 3, 3, 3, 4, 3, 3, 4, 4, 4, 4, 4, 4, 4,
       3, 3, 3, 4, 4, 3, 4, 4, 3, 4, 3, 4, 4, 4, 2, 2, 2, 3, 3, 3, 4, 3,
       3, 3, 4])

The above shows the predicted score for the input essay from the testing dataset

### Confusion matrix and Accuracy Score

Using the <b>accuracy_score function</b> from sklearn library to generate accuracy for our SVC model. <br>
The formula for the accuracy score is <b>Accuracy = (TP+TN)/(TP+TN+FP+FN)</b><br>
TP = True Positive <br>
TN = True Negative<br>
FP = False Positive <br>
FN = False Negative<br>

In [31]:
from sklearn.metrics import accuracy_score
print("Accuracy = " + str(accuracy_score(y_test,y_pred)))

Accuracy = 0.7265917602996255


The model has an Accuracy of approximately 0.73 (for the testing dataset)

<b>Confusion matrix</b><br>
A confusion matrix summarises the prediction results of the model. It outputs the number of correct predictions of each class and the number of incorrect predictions of each class in the form of a square matrix. The number of rows and columns in a confusion matrix depends on the number of classes. <br>
For example, since our data has 6 classes, our confusion matrix is a 6x6 matrix. Each row of the matrix corresponds to an actual class, while each column of the matrix corresponds to a predicted class.

In [32]:
from sklearn.metrics import confusion_matrix 
cm = confusion_matrix(y_test, y_pred)
cm

array([[ 0,  2,  0,  0,  0,  0],
       [ 0, 12,  7,  1,  0,  0],
       [ 0,  1, 91, 23,  0,  0],
       [ 0,  0, 28, 91,  0,  0],
       [ 0,  0,  1,  9,  0,  0],
       [ 0,  0,  0,  1,  0,  0]])

From the confusion matrix printed above we can gather information about number of True Positive, True Negative, False Positive and False negative for each class (6 classes, which is score of 1 to 6)

### Quadratic Weighted Kappa (QWK)

<b>Quadratic Weighted Kappa</b> is calculated using values between the actual/expected values and the predicted value. The metric varies from 0 to 1. A value nearer to 1 indicates that prediction is close to the actual value. 

It is useful because it takes into consideration how far away the prediction is from the actual. It does this by creating a weight matrix, where the prediction that is further away from the actual will receive a lower score as compared to prediction that is nearer to the actual values. <br>
Whereas for accuracy, when the actual is 2, it considers the prediction of 1 and a prediction of 1000 as equal. It counts how many prediction is true/false without considering how far away are these predictions are from the actual values. <br>

Since the Quadratic Kappa Metric is the same as cohen kappa metric in Sci-kit learn when weights are set to 'Quadratic', I will be using the cohen kappa metric code and try to aim for a score > 0.6.

In [33]:
import sklearn
sklearn.metrics.cohen_kappa_score(y_test, y_pred, weights = "quadratic")

0.6565754799849417

Here we can see that the QWK score for our model is approximately 0.66

***
## Kaggle prediction and submission

We will read the input from FIT1043-Essay-Features-Submission.csv and predict the essay score using our model.

In [34]:
#read the data from FIT1043-Essay-Features-Submission.csv file
kaggle = pd.read_csv('FIT1043-Essay-Features-Submission.csv')
#Extract the features we are using
kaggle = kaggle.iloc[:,[1,2,3,7,10,12,16,17]].values

#Scaling
kaggle = sc.transform(kaggle)
#Predict the score
kaggle_test = clf.predict(kaggle)

In [35]:
kaggle_test

array([4, 3, 3, 4, 4, 4, 3, 3, 3, 3, 3, 3, 4, 3, 4, 4, 4, 4, 3, 3, 3, 3,
       4, 4, 4, 4, 4, 4, 4, 3, 2, 4, 3, 3, 4, 3, 4, 4, 3, 3, 3, 3, 3, 3,
       2, 3, 3, 4, 4, 3, 3, 4, 4, 4, 3, 4, 3, 3, 4, 4, 2, 3, 3, 4, 3, 3,
       4, 4, 3, 4, 4, 3, 3, 3, 3, 3, 3, 3, 4, 4, 4, 3, 4, 3, 4, 2, 4, 4,
       2, 3, 3, 3, 3, 3, 4, 3, 4, 4, 3, 3, 3, 4, 2, 4, 3, 4, 3, 3, 4, 4,
       4, 3, 4, 4, 4, 3, 3, 4, 2, 3, 4, 4, 4, 3, 3, 4, 3, 4, 3, 2, 4, 4,
       2, 3, 4, 3, 3, 4, 2, 3, 4, 4, 4, 3, 3, 4, 4, 3, 4, 4, 4, 3, 4, 3,
       4, 3, 3, 3, 3, 4, 4, 3, 2, 3, 4, 4, 3, 3, 2, 4, 3, 4, 4, 3, 4, 3,
       4, 3, 3, 4, 4, 3, 4, 4, 4, 4, 3, 3, 3, 3, 3, 4, 3, 4, 3, 4, 4, 3,
       3])

The above shows the predicted essay score for the kaggle submission

The predicted essay score will be added to the 30856183-Jing-Wei-Loh-1.csv file

In [36]:
submission['score'] = kaggle_test

In [37]:
submission.to_csv('30856183-Jing-Wei-Loh-1.csv', index=False)

***
## Conclusion

The prediction model that we have created is fairly accurate, with accuracy of approximately 0.70 and QWK score of approximately 0.63